linera-service/src/proxy/grpc.rs

```diff
@@ -468,6 +470,59 @@ where
         Ok(Response::new(certificate.try_into()?))
     }
 
+    #[instrument(skip_all, err(Display))]
+    async fn download_certificates(
+        &self,
+        request: Request<CertificatesBatchRequest>,
+    ) -> Result<Response<CertificatesBatchResponse>, Status> {
+        let hashes = request
+            .into_inner()
+            .hashes
+            .into_iter()
+            .map(linera_base::crypto::CryptoHash::try_from)
+            .collect::<Result<Vec<linera_base::crypto::CryptoHash>, _>>()?;
+        let mut certificates: Vec<linera_chain::data_types::Certificate> = self
+            .0
+            .storage
+            .read_certificates(hashes)
+            .await
+            .map_err(|err| Status::from_error(Box::new(err)))?
+            .into_iter()
+            .collect::<Vec<_>>();
+
+        // Tries to serialize the certificates into a proto message.
+        // Returns `true` when encoding succeeds and `false` when it fails.
+        // Returns error when conversion to protobuf message fails.
+        fn try_proto_serialize<T, U>(c: T) -> Result<bool, Status>
+        where
+            U: Message + TryFrom<T, Error = GrpcProtoConversionError>,
+        {
+            // Use 70% of the max message size as a buffer capacity.
+            // Leave 30% as overhead.
+            let message_capacity = GRPC_MAX_MESSAGE_SIZE as f64 * 0.7;
+            let mut buff: Vec<u8> = Vec::with_capacity(message_capacity as usize);
+            let grpc_response = U::try_from(c)?;
+            Ok(prost::Message::encode(&grpc_response, &mut buff).is_ok())
+        }
+
+        // Reverse the order to drop from the back.
+        certificates.reverse();
```

Comment:
This is wrong - `.pop()` removes _last_ element so no need to reverse. Will fix but don't want to stop the CI right now.

Reply:
Fixed.